src/NuGet.Core/NuGet.Protocol/Resources/VulnerabilityInfoResourceV3.cs

```diff
@@ -0,0 +1,168 @@
+                return serviceIndex.GetServiceEntryUri(ServiceTypes.VulnerabilityInfo);
+            }
+        }
+
+        internal async Task<IReadOnlyDictionary<string, IReadOnlyList<VulnerabilityInfo>>> GetVulnerabilityDataAsync(
+            Uri uri,
+            string name,
+            SourceCacheContext cacheContext,
+            ILogger logger,
+            CancellationToken cancellationToken)
+        {
+            HttpSourceResource httpSourceResource = await _sourceRepository.GetResourceAsync<HttpSourceResource>(cancellationToken);
+
+            HttpSourceCacheContext httpSourceCacheContext = HttpSourceCacheContext.Create(cacheContext, isFirstAttempt: true);
+            var request = new HttpSourceCachedRequest(uri.OriginalString, "vuln_data_" + name, httpSourceCacheContext);
+            IReadOnlyDictionary<string, IReadOnlyList<VulnerabilityInfo>>? data = await httpSourceResource.HttpSource.GetAsync(request,
+                result =>
+                {
+                    using (var textReader = new StreamReader(result.Stream))
+                    using (var jsonReader = new JsonTextReader(textReader))
+                    {
+                        var parsed = JsonExtensions.JsonObjectSerializer.Deserialize<IReadOnlyDictionary<string, IReadOnlyList<VulnerabilityInfo>>>(jsonReader);
+                        return Task.FromResult(parsed);
+                    }
+                },
+                logger,
+                cancellationToken);
+
+            if (data == null)
+            {
+                string message = "Package source '{0}' vulnerability data file '{1}' ({2}) did not successfully deserialize. This indicates a problem with the JSON file contents returned from the server.";
+                throw new FatalProtocolException(message);
+            }
+
+            return data;
+        }
+
+        public async Task<GetVulnerabilityInfoResult> GetVulnerabilityInfoAsync(SourceCacheContext cacheContext, ILogger logger, CancellationToken cancellationToken)
+        {
+            List<Exception>? exceptions = null;
+            List<IReadOnlyDictionary<string, IReadOnlyList<VulnerabilityInfo>>>? knownVulnerabilities = null;
+
+            try
+            {
+                IReadOnlyList<V3VulnerabilityIndexEntry> indexEntries = await GetVulnerabilityFilesAsync(cacheContext, logger, cancellationToken);
+
+                if (indexEntries.Count == 0)
+                {
+                    knownVulnerabilities = new();
+                    return new GetVulnerabilityInfoResult(knownVulnerabilities, exceptions);
+                }
+
+                var tasks = new Task<IReadOnlyDictionary<string, IReadOnlyList<VulnerabilityInfo>>>[indexEntries.Count];
+                for (int i = 0; i < tasks.Length; i++)
+                {
+                    V3VulnerabilityIndexEntry indexEntry = indexEntries[i];
+                    tasks[i] = GetVulnerabilityDataAsync(indexEntry, cacheContext, logger, cancellationToken);
```

Comment:
I would clamp this to an upper bound to avoid flooding the thread pool with tasks if nuget.org ever decides to have thousands of files.

Reply:
I think any `WhenAll` http communication should have an upper bound. This one is a sneaky/unobvious one.

Reply:
I put in a basic limit of 16 for now. A better throttle so that the max limit isn't all done concurrently can be a perf optimization later.

Reply:
My comment was about the parallelization degree instead of the total count of pages actually.